src/libraries/System.IO/tests/BinaryWriter/BinaryWriter.EncodingTests.cs

```diff
@@ -0,0 +1,236 @@
+        {
+            string stringToWrite = new string('\u2023', stringLengthInChars); // TRIANGULAR BULLET
+            byte[] expectedBytes = Encoding.UTF8.GetBytes(stringToWrite);
+
+            MemoryStream stream = new MemoryStream();
+            BinaryWriter writer = new BinaryWriter(stream);
+
+            writer.Write(stringToWrite);
+            stream.Position = 0;
+
+            Assert.Equal(expectedBytes.Length /* byte count */, new BinaryReader(stream).Read7BitEncodedInt());
+            Assert.Equal(expectedBytes, stream.GetBuffer()[Get7BitEncodedIntByteLength((uint)expectedBytes.Length)..(int)stream.Length]);
+        }
+
+        [Theory]
+        [InlineData(128 * 1024)]
+        [InlineData(768 * 1024)]
+        [InlineData(2 * 1024 * 1024)]
+        public void WriteString_NotUtf8(int stringLengthInChars)
+        {
+            string stringToWrite = GenerateLargeUnicodeString(stringLengthInChars);
+            byte[] expectedBytes = Encoding.Unicode.GetBytes(stringToWrite);
+
+            MemoryStream stream = new MemoryStream();
+            BinaryWriter writer = new BinaryWriter(stream, Encoding.Unicode /* little endian */);
+
+            writer.Write(stringToWrite);
+            stream.Position = 0;
+
+            Assert.Equal(expectedBytes.Length /* byte count */, new BinaryReader(stream).Read7BitEncodedInt());
+            Assert.Equal(expectedBytes, stream.GetBuffer()[Get7BitEncodedIntByteLength((uint)expectedBytes.Length)..(int)stream.Length]);
+        }
+
+        [Fact]
+        public unsafe void WriteChars_VeryLargeArray_DoesNotOverflow()
+        {
+            const nuint INPUT_LEN_IN_CHARS = 1_500_000_000;
+            const nuint OUTPUT_LEN_IN_BYTES = 3_500_000_000; // overallocate
+
+            SafeBuffer unmanagedInputBuffer;
+            SafeBuffer unmanagedOutputBufer;
+            try
+            {
+                unmanagedInputBuffer = SafeBufferUtil.CreateSafeBuffer(INPUT_LEN_IN_CHARS * sizeof(char));
+                unmanagedOutputBufer = SafeBufferUtil.CreateSafeBuffer(OUTPUT_LEN_IN_BYTES * sizeof(byte));
+            }
+            catch (OutOfMemoryException)
+            {
+                return; // skip test in low-mem conditions
+            }
+
+            using (unmanagedInputBuffer)
+            using (unmanagedOutputBufer)
+            {
+                Span<char> inputSpan = new Span<char>((char*)unmanagedInputBuffer.DangerousGetHandle(), (int)INPUT_LEN_IN_CHARS);
+                inputSpan.Fill('\u0224'); // LATIN CAPITAL LETTER Z WITH HOOK
+                Stream outStream = new UnmanagedMemoryStream(unmanagedOutputBufer, 0, (long)unmanagedOutputBufer.ByteLength, FileAccess.ReadWrite);
+                BinaryWriter writer = new BinaryWriter(outStream);
+
+                writer.Write(inputSpan); // will write 3 billion bytes to the output
+
+                Assert.Equal(3_000_000_000, outStream.Position);
+            }
+        }
```

Comment:
thank you for writing all the tests! and especially covering this particular edge case! 👍